File: Contest Volumes/Volume 120 (12000-12099)/UVa_12068_Harmonic_Mean.java

// 12068 - Harmonic Mean
// This solution has not been submitted yet.

import java.util.Scanner;
import java.math.BigInteger;

class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int S = sc.nextInt();
		for (int s = 1; s <= S; s++) {
			int N = sc.nextInt();
			BigInteger[] numbers = new BigInteger[N];
			for (int i = 0; i < N; i++)
				numbers[i] = sc.nextBigInteger();
			BigInteger numerator = BigInteger.valueOf(N), denominator = BigInteger.ZERO;
			for (int i = 0; i < N; i++) {
				numerator = numerator.multiply(numbers[i]);
				BigInteger b = BigInteger.ONE;
				for (int j = 0; j < N - 1; j++) {
					b = b.multiply(numbers[(i + j) % N]);
				}
				denominator = denominator.add(b);
			}
			BigInteger g = numerator.gcd(denominator);
			System.out.println("Case " + s + ": " + numerator.divide(g) + "/" + denominator.divide(g));
		}
	}
}

File: Contest Volumes/Volume 120 (12000-12099)/UVa_12068_Harmonic_Mean.cpp

/*
	UVa 12068 - Harmonic Mean

	To build using Visual Studio 2012:
		cl -EHsc -O2 UVa_12068_Harmonic_Mean.cpp

	This is an accepted solution.
*/

#include <cstdio>

long long gcd(long long x, long long y)
{
	if (x < y)
		return gcd(y, x);
	else
	    return y == 0 ? x : gcd(y, x % y);
}

int main()
{
	int S;
	scanf("%d", &S);
	for (int s = 1; s <= S; s++) {
		const int N_max = 8;
		int N, numbers[N_max];
		scanf("%d", &N);
		for (int i = 0; i < N; i++)
			scanf("%d", &numbers[i]);
		long long numerator = N, denominator = 0;
		for (int i = 0; i < N; i++) {
			numerator *= numbers[i];
			long long d = 1;
			for (int j = 0; j < N - 1; j++)
				d *= numbers[(i + j) % N];
			denominator += d;
		}
		long long g = gcd(numerator, denominator);
		printf("Case %d: %lld/%lld\n", s, numerator / g, denominator / g);
	}
	return 0;
}

